src/runtime/virtcontainers/nydusd.go

```diff
@@ -0,0 +1,424 @@
+	fields := logrus.Fields{
+		"path": nd.path,
+		"args": strings.Join(args, " "),
+	}
+	nd.Logger().WithFields(fields).Info()
+	if err := nd.startFn(cmd); err != nil {
+		return pid, err
+	}
+	// Monitor nydusd's stdout/stderr and stop sandbox if nydusd quits
+	go func() {
+		scanner := bufio.NewScanner(r)
+		for scanner.Scan() {
+			nd.Logger().Info(scanner.Text())
+		}
+		nd.Logger().Info("nydusd quits")
+		// Wait to release resources of nydusd process
+		_, err = cmd.Process.Wait()
+		if err != nil {
+			nd.Logger().WithError(err).Warn("nydusd quits")
+		}
+		if onQuit != nil {
+			onQuit()
+		}
+	}()
+	if err := nd.setupShareDirFn(); err != nil {
+		return pid, err
+	}
+	nd.pid = cmd.Process.Pid
+	return nd.pid, nil
+}
+
+func (nd *nydusd) args() ([]string, error) {
+	logLevel := "info"
+	if nd.debug {
+		logLevel = "debug"
+	}
+	args := []string{
+		"--log-level", logLevel,
+		"--apisock", nd.apiSockPath,
+		"--sock", nd.sockPath,
+	}
+	if len(nd.extraArgs) > 0 {
+		args = append(args, nd.extraArgs...)
+	}
+	return args, nil
+}
+
+func (nd *nydusd) valid() error {
+	if len(nd.path) == 0 {
+		return errNydusdDaemonPathEmpty
+	}
+	if len(nd.sockPath) == 0 {
+		return errNydusdSocketPathEmpty
+	}
+	if len(nd.apiSockPath) == 0 {
+		return errNydusdAPISocketPathEmpty
+	}
+	if len(nd.sourcePath) == 0 {
+		return errNydusdSourcePathEmpty
+	}
+	if _, err := os.Stat(nd.sourcePath); err != nil {
+		return errNydusdSourceNotAvailable
+	}
+	return nil
+}
+
+func (nd *nydusd) setupPassthroughFS() error {
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("from", nd.sourcePath).
+		WithField("dest", sharedPathInGuest).Info("prepare mount passthroughfs")
+
+	mr := NewMountRequest(nydusPassthroughfsType, nd.sourcePath, "")
+	return nc.Mount(sharedPathInGuest, mr)
+}
+
+func (nd *nydusd) Mount(opt MountOption) error {
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("from", opt.source).
+		WithField("dest", opt.mountpoint).Info("prepare mount rafs")
+
+	mr := NewMountRequest(nydusRafsType, opt.source, opt.config)
+	return nc.Mount(opt.mountpoint, mr)
+}
+
+func (nd *nydusd) Umount(mountpoint string) error {
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("mountpoint", mountpoint).Info("umount rafs")
+	return nc.Umount(mountpoint)
+}
+
+func (nd *nydusd) Stop(ctx context.Context) error {
+	if err := nd.kill(ctx); err != nil {
+		nd.Logger().WithError(err).WithField("pid", nd.pid).Warn("kill nydusd failed")
+		return nil
+	}
+
+	err := os.Remove(nd.sockPath)
+	if err != nil {
+		nd.Logger().WithError(err).WithField("path", nd.sockPath).Warn("removing nydusd socket failed")
+	}
+	err = os.Remove(nd.apiSockPath)
+	if err != nil {
+		nd.Logger().WithError(err).WithField("path", nd.apiSockPath).Warn("removing nydusd api socket failed")
+	}
+	return nil
+}
+
+func (nd *nydusd) Logger() *logrus.Entry {
```

Comment:
This lookg wrong to me. Please see https://github.com/kata-containers/kata-containers/pull/326 for an example of the expected `SetLogger()` API and how to call it.

Reply:
here, I refer to [virtiofsd Logger](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/virtiofsd.go#L218), virtiofsd and nydusd are hypervisor subsystems, so I think it is ok？

And I see, hvLogger has done the [logger setting](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/hypervisor.go#L149)